### ISSUE TITLE
Proposal: Diffusion decoder with flow matching as optional trajectory planner

### ISSUE THREAD

**Issue by m-zain-khawaja** [autowarefoundation/auto_e2e#45]
## Summary

Unlike discrete textual tokens or autoregressive sequence decoding (which can result in jerky, computationally inefficient trajectory predictions), flow matching operates over continuous values directly producing a smooth continuous motion output - as implemented in the Alpamayo model. This proposal seeks to add a swappable module in the trajectory planner to use flow matching methodology.

## Key Benefits

- Continuous Trajectories: It defines a probability path (a velocity field) connecting noise to the final trajectory plan.
- Geometry-Aware Control: By bypassing discrete motion tokens, the decoder preserves scalar geometry and kinematic constraints natively.
- Real-time Latency: The optimization streamlines both training and inference, achieving single-digit millisecond inference times necessary for safety
- Continuous Motion Output: The flow-matching head regresses the velocity vectors to map the straightest possible path from the starting state to the optimal, multi-modal future trajectory.

## Implementation Proposal
To integrate a flow-matching based trajectory decoder into a Bird's-Eye-View (BEV) feature module, we couldt transform the 2D spatially grid-aligned BEV features into an information-dense conditional context vector. This vector would then guide a continuous Vector Field (Velocity Field) Regressor to denoise raw gaussian noise into smooth, kinematically sound coordinates.

### Architectural Changes
```
BEV Feature Map] ────> [Conv2D / Flatten] ────┐
(B, C, H, W)                                  │
                                              ▼
[Noisy Actions u_t] ──> [Action Projection] ──> [Cross-Attention Layer] ──> [AdaLN (Time t)] ──> [Velocity Field Output v_θ]
(B, T, 2)                                                                                       (B, T, 2)

```
```python
class TimestepEmbedding(nn.Module):
    """Maps scalar timesteps t to high-dimensional sinusoidal embeddings."""
    def __init__(self, dim):
        super().__init__()
        self.dim = dim

    def forward(self, t):
        device = t.device
        half_dim = self.dim // 2
        embeddings = math.log(10000) / (half_dim - 1)
        embeddings = torch.exp(torch.arange(half_dim, device=device) * -embeddings)
        embeddings = t[:, None] * embeddings[None, :]
        embeddings = torch.cat((embeddings.sin(), embeddings.cos()), dim=-1)
        return embeddings

class FlowMatchingControlDecoder(nn.Module):
    def __init__(self, bev_channels=256, embed_dim=256, traj_len=64, num_heads=4):
        super().__init__()
        self.traj_len = traj_len
        self.embed_dim = embed_dim
        
        # Time Step Embedding Generator
        self.time_mlp = nn.Sequential(
            TimestepEmbedding(embed_dim),
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, embed_dim)
        )
        
        # Action Input Projection (Maps continuous [acceleration, curvature] pairs)
        self.action_proj = nn.Linear(2, embed_dim)
        
        # BEV Image Feature Processing 
        # Compresses the spatial map or downsamples if map resolution is massive
        self.bev_downsample = nn.Sequential(
            nn.Conv2d(bev_channels, embed_dim, kernel_size=3, stride=1, padding=1),
            nn.GroupNorm(32, embed_dim),
            nn.SiLU()
        )
        
        # Attention Layer: Action queries look over the entire spatial BEV image map
        self.cross_attn = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
        
        # Layer Normalization Blocks
        self.norm_bev = nn.LayerNorm(embed_dim)
        self.norm_action = nn.LayerNorm(embed_dim)
        
        # Adaptive Layer Normalization (AdaLN) for condition scaling via time injection
        self.adaln_gamma = nn.Linear(embed_dim, embed_dim)
        self.adaln_beta = nn.Linear(embed_dim, embed_dim)
        
        # Prediction Output Head: Regresses the velocity field vectors (da/dt, d_kappa/dt)
        self.velocity_head = nn.Sequential(
            nn.Linear(embed_dim, embed_dim),
            nn.SiLU(),
            nn.Linear(embed_dim, 2)
        )

    def forward(self, u_t, t, bev_features):
        """
        u_t: Noisy control actions at flow time step t. Shape: (B, Traj_Len, 2) -> [accel, curvature]
        t: Continuous flow-matching time step. Shape: (B,) scalar tensor between [0.0, 1.0]
        bev_features: Image feature tensor from your BEV module backbone. Shape: (B, C, H, W)
        """
        B, _, H, W = bev_features.shape
        
        # 1. Process and encode continuous flow time
        t_embed = self.time_mlp(t) # Shape: (B, embed_dim)
        
        # 2. Flatten 2D BEV features into an attention-ready sequence
        processed_bev = self.bev_downsample(bev_features) # Shape: (B, embed_dim, H, W)
        bev_flat = processed_bev.permute(0, 2, 3, 1).reshape(B, H * W, self.embed_dim)
        bev_context = self.norm_bev(bev_flat) # Shape: (B, H*W, embed_dim)
        
        # 3. Project continuous action trajectories to Query space
        action_query = self.norm_action(self.action_proj(u_t)) # Shape: (B, Traj_Len, embed_dim)
        
        # 4. Perform cross-attention feature fusion
        # Queries (Actions) search through Keys & Values (BEV Context)
        attn_out, _ = self.cross_attn(query=action_query, key=bev_context, value=bev_context)
        
        # 5. Inject time embeddings using AdaLN modulation
        gamma = self.adaln_gamma(t_embed).unsqueeze(1) # Shape: (B, 1, embed_dim)
        beta = self.adaln_beta(t_embed).unsqueeze(1)   # Shape: (B, 1, embed_dim)
        fused_features = attn_out * (1 + gamma) + beta
        
        # 6. Regress Vector Field Velocity
        v_theta = self.velocity_head(fused_features) # Shape: (B, Traj_Len, 2)
        return v_theta

**Comment by riita10069** [comment]
Thanks for this proposal — flow matching is a compelling direction for the trajectory decoder. I've been looking into the implementation patterns from Alpamayo (NVIDIA), torchcfm (Meta), GoalFlow (CVPR 2025), and Diffusion Policy to understand how best to integrate this.

## Proposed Implementation Plan

### 1. Swappable planner via registry pattern (same as view fusion)

```
Model/model_components/trajectory_planning/
├── __init__.py              ← PLANNER_REGISTRY + build_planner()
├── base.py                  ← BasePlanner interface
├── gru_planner.py           ← Current TrajectoryPlanner (renamed)
└── flow_matching_planner.py ← New FlowMatchingDecoder
```

`AutoE2E` would accept `planner_mode="gru"` or `planner_mode="flow_matching"`, keeping both available for comparison and ablation.

### 2. Loss computation: external, following Alpamayo pattern

Looking at how published implementations handle this:
- **Alpamayo** (NVIDIA): model predicts velocity field, separate `compute_loss_from_pred(pred, x, noise)` computes `MSE(pred, x - noise)` externally
- **torchcfm** (Meta): `loss = torch.mean((vt - ut) ** 2)` in user training loop
- **GoalFlow** (CVPR 2025): uses L1 instead of MSE for stability: `F.l1_loss(pred_trajectory, target)`
- **Diffusion Policy** (Stanford): separate `compute_loss(batch)` and `predict_action(obs_dict)` methods

The consensus is: **model returns velocity prediction, loss is computed externally.** This keeps the model's forward() clean and composable.

### 3. Training vs inference API (Alpamayo-style separation)

```python
# Training loop (external):
u_t, t, target_velocity = planner.construct_training_data(trajectory_target)
v_theta, ego_hidden = planner(bev_features, visual_history, egomotion_history,
                               mode="train", noisy_trajectory=u_t, flow_timestep=t)
loss = F.mse_loss(v_theta, target_velocity)

# Inference:
trajectory, ego_hidden = planner(bev_features, visual_history, egomotion_history,
                                  mode="infer")
# Internally runs Euler ODE integration for num_inference_steps
```

This follows Alpamayo's pattern where `construct_training_data()` is called before the forward pass to generate the noisy interpolants and target velocities. The random noise is generated outside the model forward, making the loss computation fully transparent.

### 4. BasePlanner interface

Both planners return `(trajectory, ego_hidden)`:
- **GRU mode**: trajectory = direct prediction `(B, 128)`, ego_hidden = GRU final state `(B, 256)`
- **Flow Matching train mode**: trajectory = v_theta `(B, 128)`, ego_hidden = conditioning vector `(B, 256)`
- **Flow Matching infer mode**: trajectory = denoised result after ODE `(B, 128)`, ego_hidden = conditioning vector `(B, 256)`

ego_hidden is passed to FutureState regardless of planner mode.

## Questions

1. **Loss type**: Alpamayo uses MSE, GoalFlow uses L1 (arguing it's more stable for trajectory prediction). Do you have a preference? We could make this configurable like our existing `TrajectoryImitationLoss`.

2. **Number of inference steps**: Alpamayo uses 10 Euler steps. Is this the target, or should we also support adaptive ODE solvers (e.g., dopri5) for higher accuracy at the cost of variable step count?

3. **Classifier-free guidance**: Alpamayo uses 10% guidance dropout during training to enable classifier-free guidance at inference. Should we include this from the start, or add it as a follow-up?

4. **Conditioning**: Your proposal uses BEV features as the only conditioning context for cross-attention. Should egomotion_history and visual_history also condition the velocity field prediction (e.g., via AdaLN alongside the timestep embedding), or is BEV context sufficient since it already encodes the scene?

Happy to start implementation once we align on these points.

**Comment by m-zain-khawaja** [comment]
Thank you @riita10069 this is great - regarding specific questions:

- For Loss-type, Inference steps and Classifier-free guidance, could we please follow the same approach as Alpamayo
- I think it would be useful to condition the velocity field prediction based on ego_motion history and visual_history - are you able to please reflect this change in your PR

**Comment by riita10069** [comment]
Thanks for the detailed proposal — I've implemented this as a swappable flow-matching planner in PR #51, following the architecture you outlined (action projection -> cross-attention over BEV -> AdaLN time conditioning -> velocity field), and aligned it with Alpamayo on the points below. It's now rebased on latest main with CI green.

On loss-type: the planner trains with the conditional flow-matching velocity MSE in FlowMatchingPlanner.compute_planner_loss — F.mse_loss(v_θ(u_t, t, c), x₁ − x₀) on the linear-OT path u_t = (1−t)·x₀ + t·x₁, which matches Alpamayo's mse(pred, x − noise). The flow-timestep sampler also follows Alpamayo's default: a shifted Beta(1.5, 1.0) schedule (t = 0.999 − 0.999 · Beta(1.5, 1).sample()) that biases t toward the noisy end, with timestep_sampler="uniform" kept as an opt-out. (Ref: Lipman et al., 2022, arXiv:2210.02747.)

On inference steps: inference integrates dx/dt = v_θ with a fixed-step Euler solver over t in [0, 1], default num_inference_steps=10, the same as Alpamayo's num_inference_steps=10 with int_method="euler". Happy to change the default if you'd prefer a different budget.

On conditioning the velocity field on ego-motion and visual history: this is in place. _modulation_conditioning() projects both visual_history and egomotion_history and feeds them, together with the Fourier time embedding, into the AdaLN modulation that scales and shifts the velocity field, while the BEV map enters via cross-attention. So v_θ attends to BEV and is modulated by ego-motion, visual history, and flow time. Glad to switch to a different injection (for example in-context concatenation via a KV-cache, as Alpamayo does) if you'd prefer.

On classifier-free guidance: I held off on this one. In the released Alpamayo code the guidance hooks look present but inert — construct_training_data sets is_drop_guidance=None with no condition-dropout, and inference_guidance_weight / train_ignore_guidance_rate aren't read during training or sampling, and the paper doesn't discuss CFG. So matching Alpamayo here effectively means running without it, which is what the PR does (a single velocity evaluation per Euler step). If you'd like CFG as a follow-up, I'd be glad to align on what we'd guide toward — in driving the guidance target is often a cost or constraint (cf. Diffusion Planner, arXiv:2501.15564) rather than a dropped categorical condition, and it roughly doubles inference cost (cond + uncond per step), so it might fit best as a separate issue or PR. Canonical reference: Ho & Salimans, 2022, arXiv:2207.12598.

**Comment by m-zain-khawaja** [comment]
@riita10069 - thanks for the detailed explantion - sounds great!
